---
title: "Testing *cn.mops* using TIL01 as reference"
author: "Simon Renny-Byfield"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

Jeff mentioned that I should examine the cnv calls of [cn.ops](http://www.bioinf.jku.at/software/cnmops/) using TIL01 and TIL11 which already have cnvs called using a chip, detailed in a [Swanson-Wagner](http://www.ncbi.nlm.nih.gov/pubmed/21036921) paper. The issue I can foresee is that the analysis was done with the maize reference version 2, whereas we are working with maize ref3... Perhaps this means having to convert co-ordinates, or revert back to maize ref2 to assess the hit rate of cn.mops.

##Mapping the TIL data using BWA-MEM##

The data for TIL01 and TIL11 can be found in `/group/jrigrp4/cn.mops/TIL_data` on farm. 

The first step is to generate index files for the refercence genome. The reference in question is stored in `/group/jrigrp4/teosinte_parents/genomes` and I used the following command to generate the index:
```
screen bwa index -a bwtsw Zea_mays.AGPv3.22.dna.genome.fa
```
Following this the data are mapped using BWA-MEM and a batch script:
```
!/bin/bash
#OUTDIR=/group/jrigrp4/cn.mops/output
#SBATCH -D /group/jrigrp4/cn.mops/TIL_data
#SBATCH -o /group/jrigrp4/cn.mops/logs/cn.mops_out_log-%j.txt
#SBATCH -e /group/jrigrp4/cn.mops/logs/cn.mops_err_log-%j.txt
#SBATCH -J bwa
#SBATCH --mem=10000
#SBATCH --cpus-per-task 10


echo "Starting bwa job: "
date

cmd="bwa mem -M -t 10 -v 1 -p /group/jrigrp4/teosinte_parents/genomes/Zea_mays.AGPv3.22.dna.genome.fa TIL01_3510_3807_3510_N_TIP521_4_R1.fastq TIL01_3510_3807_3510_N_TIP521_4_R2.fastq | samtools view -Sb - > TIL01.bam"
echo $cmd
eval $cmd

echo "Job Ending: "
date
```
