---
title: "Running cn.mops on 20 teosinte from Palmar Chico"
author: "Simon Renny-Byfield"
date: "`r Sys.Date()`"
output: html_document
---

This document records efforts to run the `R` package [`cn.mops`](http://www.bioconductor.org/packages/release/bioc/html/cn.mops.html) on 20 teosinte individuals from the Palmar Chico population, Mexico. As input to the program I have sorted `.bam` files (mapped by Vince Buffalo using [paap.py](https://github.com/vsbuffalo/paap)). These files symlinked in `/group/jrigrp4/cn.mops/data` from another location `/group/jrigrp4/teosinte-parents/aln/`.

Working in the `/group/jrigrp4/cn.mops/` directory I indexed the appropriate `.bam` files using the following scrip `scripts/index.bam.sh`:
```
#!/bin/bash -l
#!/bin/bash
#OUTDIR=/group/jrigrp4/cn.mops/output
#SBATCH -D /group/jrigrp4/cn.mops/output
#SBATCH -o /group/jrigrp4/cn.mops/logs/index_out_log-%j.txt
#SBATCH -e /group/jrigrp4/cn.mops/logs/index_err_log-%j.txt
#SBATCH -J index
#SBATCH --array=0-19
#SBATCH --mem-per-cpu=8000

##Simon Renny-Byfield, UC Davis, December 16 2014


echo "Starting Job:"
date

files=($(ls -d ../data/*sorted.bam))

#now index each .bam file

cmd="samtools index ${files[$SLURM_ARRAY_TASK_ID]}"
echo $cmd
eval $cmd

echo "End Job: "
date
```
Follwoing this I modified a script the Jinliang gave me. This script reads in the bam files, assess coverage and outputs `.bed` style output for CNVs and PAVs. The script is saved as `scripts/cnv_cnmpos.R`. At first I limited the analysis to the first 3 `.bam` files in the list. This is done by modifying the bam files list like this `BAMFiles[1:3]`.
```
# Simon Renny-Byfield
# 16/12/2014
# do a CNV call for 20 teosine lines from Palmar Chico
# modified from a script provided by Jinliang

#source("http://bioconductor.org/biocLite.R")
#biocLite("cn.mops")
len<-100
library(cn.mops)
BAMFiles <- list.files(path="/group/jrigrp4/cn.mops/data", pattern="sorted.bam$")
setwd("/group/jrigrp4/cn.mops/data")

chrs <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")
#for each chromosome
for (i in chrs ) {
        setwd("/group/jrigrp4/cn.mops/data")
        #try just chr 10 for now
        #chrs<-10
        #print(BAMFiles)
        bamDataRanges <- getReadCountsFromBAM(BAMFiles, refSeqName=i,  mode="paired",WL=len)

        #resHaplo <- haplocn.mops(bamDataRanges)
        #resCN <- calcIntegerCopyNumbers(resHaplo)
        # This function performs the cn.mops algorithm for copy number detection in NGS data
        res <- cn.mops(bamDataRanges)
        resCNV <- calcIntegerCopyNumbers(res)

        ### transform GRanges to data.frame
        mycnv <- cnvr(resCNV)


        cnvdf <- data.frame(chr=as.character(seqnames(mycnv)), start=start(mycnv), end=end(mycnv))
        callcnv <- mcols(mycnv)
        cnvdf <- cbind(cnvdf, callcnv)
        setwd("/group/jrigrp4/cn.mops/output")
        save(file=paste0("bamDataRanges_", i, ".RData"), list=c("bamDataRanges","cnvdf") )
}#for
```
Importantly this must be run from insde a slurm script. I used the following to get the job done:
```
#!/bin/bash
#OUTDIR=/group/jrigrp4/cn.mops/output
#SBATCH -D /group/jrigrp4/cn.mops/output
#SBATCH -o /group/jrigrp4/cn.mops/logs/cn.mops_out_log-%j.txt
#SBATCH -e /group/jrigrp4/cn.mops/logs/cn.mops_err_log-%j.txt
#SBATCH -J cn.mops
#SBATCH --mem=24000

export R_LIBS="~/R/x86_64-pc-linux-gnu-library/3.1"

echo "Starting cn.mops job: "
date

cmd="Rscript ../scripts/cnv_cnmpos.R"
echo $cmd
eval $cmd

echo "Job Ending: "
date
```

The data are now availalbe and have been moved to my desktop in `/Users/simonrenny-byfield/CNV_PAV`. Calls from each chromosome were combined into a single file called `cnv_calls.RData`.

##Testing CNV calls using Swanson-Wagner data and TIL01##

Data from SW genome research [paper](http://genome.cshlp.org/content/20/12/1689.long#T3) is stored in `/Users/simonrenny-byfield/CNV_PAV/SW_data/SW_cnv_calls.csv`. The next step is to use this data to test our own CNV calls on TIL01. It is important to realize that some of the gene names have changed from RefV1 to RefV3.

Now we need the gene annotations to see which CNVs overlap with the genes. I can see two ways to do this:

1.  Use the untrimmed data set, that is without regard to CNVs overlapping repeats.
2.  USe the final trimmmed dataset which we will finally use in the SFS and selection analysis.

I think option 2 is probably the best so I will use the *final* set of CNVs.

Now I am just cheking iof the repeat region file is the correct one and I have the most uptodate version of the repeat gff. As it turns out AGPv3.22 is NOT the same as AGPv3.25. I have now sourced AGPv3.22 annotations and everything is on order and ready to go.

**(In)Congruence between SW ChiP and NGS data**

Of the 3805 CNV/PAV calls from the SW [paper](http://genome.cshlp.org/content/20/12/1689.long#T3), I filtered these to those called as PAVs/CNVs in `TIL01`. I then checked howm many of the genes with CNV calls actually have matched names in the AGPv3 gene set. This further limited the genes to 3155. This results in ~1000 PAVs/CNVs called in TIL01. Next we examine the overlap between calls from cn.mops and the sw chip.

```{r, echo = FALSE, eval=TRUE}
table1<-data.frame(cbind(c(152,831-152,831),c(318,2342-318,318+2024)))
colnames(table1)<-c("scored CHiP","absent CHiP")
rownames(table1)<-c("scored cn.mops","absent cn.mops","total")
print(table1)
```
<br>
<br>
`TIL01` is a highly inbred line and as such shouold be homozygous, including for any PAVs/CNVs. Worryingly there are plenty of loci, actually even more in the outbred wild Palmar Chico population, that have odd  copy numbers (i.e. 1,3, 5). After inbreeding CN should be 0,2,4,6 etc etc indicating homozygosity at each loci. 

![Integer copy number](/Users/simonrenny-byfield/CNV_PAV/output/integerdensity.png)




